src/lib/appdata.py

```diff
@@ -0,0 +1,169 @@
+        self._context = []
+        self._emitted_release = False
+        self._releases_padding = ""
+
+    @property
+    def outside_root_element(self):
+        return not self._context
+
+    @property
+    def _in_releases(self):
+        return self._context[1:] == ["releases"]
+
+    def _emit_release(self):
+        super().startElement("release", {"version": self._version, "date": self._date})
+        # TODO: add placeholder <description> if other <release> entries have one
+        super().endElement("release")
+        super().characters(self._releases_padding)
+        self._emitted_release = True
+
+    def startElement(self, name, attrs):
+        if self._in_releases and not self._emitted_release:
+            self._emit_release()
+
+        super().startElement(name, attrs)
+
+        self._context.append(name)
+
+    def characters(self, chars):
+        if self._in_releases and not self._emitted_release and chars.isspace():
+            self._releases_padding += chars
+
+        super().characters(chars)
+
+    def endElement(self, name):
+        if self._in_releases and not self._emitted_release:
+            super().characters("\n    ")
+            self._emit_release()
+            super().characters("\n  ")
+
+        self._context.pop()
+
+        if not self._context and not self._emitted_release:
+            # No <releases> found; synthesize one
+            super().characters("  ")
+            super().startElement("releases", {})
+            super().characters("\n    ")
+            self._emit_release()
+            super().characters("\n  ")
+            super().endElement("releases")
+            super().characters("\n")
+
+        super().endElement(name)
+
+
+class VerbatimLexicalHandler:
+    """Implements the poorly-documented LexicalHandler interface."""
+
+    def __init__(self, reader, stream):
+        self._reader = reader
+        self._stream = stream
+
+    def comment(self, text):
+        self._stream.write("<!--")
+        self._stream.write(text)
+        self._stream.write("-->")
+
+        # Aggravating hack alert!
+        #
+        # appdata files often include a copyright comment between the <?xml
+        # declaration and the root element. In some cases, there are two.
+        # Unfortunately, there is no representation in the SAX API for
+        # whitespace that occurs outside any element, and there's normally
+        # a newline between the comment and the opening element (or the next
+        # comment).
+        #
+        # Work around this by writing a newline in this case; this requires some
+        # coordination between the reader and this writer.
+        if self._reader.outside_root_element:
+            self._stream.write("\n")
+
+    def startCDATA(self, *args):
+        raise NotImplemented
+
+    def endCDATA(self, *args):
+        raise NotImplemented
+
+    def endDTD(self, *args):
+        raise NotImplemented
+
+
+def add_release(in_, out, version, date):
```

Comment:
I would probably s/in_/path/ here as well, or even `appdata_path`.

Reply:
Actually it doesn't have to be a path. It can be a file-like object too, and the test suite relies on this. I've made this explicit in the name.